### ISSUE TITLE
Service discovery (swarm join) unreliable, missing events

### ISSUE THREAD

**Issue by schmunk42** [docker/swarm#1405]
I am having trouble keeping a consistent node-cluster, basically all my nodes are running fine, but `swarm join` fails from time to time.

Sometimes the outage is longer than `ttl` and the swarm master removes the node. See period from `09:06:43` to `09:08:08`.
I can't see if this is a problem with my node or the discovery service (also had some `500` errors in the past days).
There's no further error, the `join` command simply doesn't happen.

```
time="2015-11-12T09:04:40Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:05:01Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:05:21Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:05:41Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:06:02Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:06:22Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:06:43Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:08:08Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:08:31Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:08:56Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:09:37Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:09:57Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:10:17Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:12:45Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
time="2015-11-12T09:13:05Z" level=info msg="Registering on the discovery service every 20s..." addr="IP:2376" discovery="token://TOKEN" 
```

Running swarm 1.0.0, docker 1.9.0 and compose 1.5.0 on EC2

**Comment by abronan** [comment]
Hi @schmunk42, there was a recent spike in usage of the token discovery which might have impacted the service. Some requests are timing out and thus nodes are getting unregistered wrongfully. We advise to use the other discovery methods in the meantime while we are investigating the root cause.

`token` is a mean to get easily started but you can get the same result using the `node` or `file` discovery without the heartbeating and removal of nodes. We recommend using the `kv` discovery for production clusters.

If you created a swarm using `docker-machine` you can first create a VM with a Consul, etcd or zookeeper instance and point to the `kv` as you would normally do (see the [docs on discovery](https://github.com/docker/swarm/tree/master/discovery)).

**Comment by schmunk42** [comment]
I haven't tried consul yet, but it looks to me that this is more a problem on the node (or swarm-agent) rather than the discovery service. I think I saw 5xx errors in other very rare cases when the discovery wasn't available.

For example, I have a green-blue redeployment script, which can be run in parallel.
When I start 1 or 2 parallel redeployments, usually there are no problems, but when I run 3 or 5 parallel redeployments, the chances for a node to disappear (running `docker info` on the master) are much higher.

I saw `SYS_NICE`, would this be a way to run the `swarm-agent` container with a higher priority?

**Comment by schmunk42** [comment]
Using consul solves my problems so far, additionally here are some stats about service discovery (recorded events from swarm-agent registering at the discovery service)

<img width="958" alt="bildschirmfoto 2015-11-19 um 16 59 45" src="https://cloud.githubusercontent.com/assets/649031/11275896/fdae4a36-8edf-11e5-8746-cbee1a187e7f.png">

Swarm agent heartbeat: 7s

First one: 4 node cluster with consul and manually triggered node changes, same amount of events from every agent reaching consul. Even during parallel redeployments of 3-5 stacks (AWS: t2.small)

Second one: 4 node swarm with token discovery without any manual changes, but many nodes can't reach the discovery service within one minute, which means they will be flagged as dead.

Third one: 2 node swarm with token discovery without any manual changes, same problems as above.

**Comment by aluzzardi** [comment]
Thanks for the report, @schmunk42.

It looks like we have problems scaling the hosted discovery /cc @vieux 

@schmunk42 Out of curiosity, what did you use to generate those metrics?

I guess we should either work on improving the hosted discovery scale or just deprecate it in favor of consul/etcd/zookeeper. Thoughts?

**Comment by auzias** [comment]
I just leave a comment to get the info about what tool did @schmunk42 to generate these.

Also I suggest, when possible, to use static file discovery. Far easier, faster and safer.
[Docker doc](https://docs.docker.com/swarm/discovery/https://docs.docker.com/swarm/discovery/#using-a-static-file-describing-the-cluster), [step-by-step tutorial using it](https://www.auzias.net/en/docker-network-multihost/#static-file-start-the-swarm-manager-while-joining-swarm-members:b286545a85a53658f9118404059fc18b)

**Comment by schmunk42** [comment]
The charts are screenshots from Kibana (ELK-stack), running a logspout in every node, tagged with the swarm name.
While the logspouts are also reporting swarm events, you can filter out these events easily in Kibana and create a visualization from them.
I'll try to create an example stack for that, since the one I have is a combination from different ones across the net.

**Comment by aluzzardi** [comment]
@schmunk42 Any issues running with consul?

As a user, would you rather use `token` for simplicity or are you happier using a consul on-prem?

**Comment by schmunk42** [comment]
Basically I wanted to use the `token` discovery, because of simplicity. I really tried to avoid consul, since it's another piece to the puzzle. But honestly, I really cursed the current `token` discovery, since I lost so many hours with it.
The reason behind is: I tried to create a very-very basic scheduler, which would detect a node failure and then reschedule the containers, but since "a node-failure" happend all the time, especially when there was a bit more load on the host, but also on very slow loads - the scheduler produced many [dead containers](https://github.com/docker/swarm/issues/1421).

On the other hand consul setup is pretty straightforward, but it needs a proper configuration with secure communication and access rights between master(s) and consul. Would be great if this could be provided in a `docker-machine` way.

My honest opinion: Note in the docs (in a yellow bow) that the current service discovery is for quick-start testing only. Or make significant improvements to it. 
With much longer timeouts, currently I'd say ... about 5 minutes, you could also work around this issue, but that is too long IMHO and also [has glitches](https://github.com/docker/machine/issues/2266).

Don't get me wrong, `swarm` is a great piece of software and it should not suffer from issues like that.

**Comment by aluzzardi** [comment]
Thank you @schmunk42, we appreciate your honest feedback.

We have some ideas in the works about having a gossip based discovery system that would have the same advantage of token (server-less) without the external dependency.

**Comment by aluzzardi** [comment]
Tagging as `doc` - we should really have this yellow box.

**Comment by thaJeztah** [comment]
Is this ready to close, now that https://github.com/docker/swarm/pull/1497 is merged, or are additional mention in the docs needed?

**Comment by abronan** [comment]
Closed by #1497
